node.go

```diff
@@ -0,0 +1,325 @@
+package amt
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/filecoin-project/go-amt-ipld/v2/internal"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+type node struct {
+	// these may both be nil if the node is empty (a root node)
+	links  [internal.Width]*link
+	values [internal.Width]*cbg.Deferred
+}
+
+var (
+	errEmptyNode      = errors.New("unexpected empty amt node")
+	errUndefinedCID   = errors.New("amt node has undefined CID")
+	errLinksAndValues = errors.New("amt node has both links and values")
+	errLeafUnexpected = errors.New("amt leaf not expected at height")
+	errLeafExpected   = errors.New("amt expected at height")
+	errInvalidCount   = errors.New("amt count does not match number of elements")
+)
+
+func newNode(nd internal.Node, allowEmpty, expectLeaf bool) (*node, error) {
+	if len(nd.Links) > 0 && len(nd.Values) > 0 {
+		return nil, errLinksAndValues
+	}
+
+	i := 0
+	n := new(node)
+	if len(nd.Values) > 0 {
+		if !expectLeaf {
+			return nil, errLeafUnexpected
+		}
+		for x := byte(0); x < internal.Width; x++ {
+			if nd.Bmap[x/8]&(1<<(x%8)) > 0 {
+				if i >= len(nd.Values) {
+					return nil, fmt.Errorf("expected at least %d values, found %d", i+1, len(nd.Values))
+				}
+				n.values[x] = nd.Values[i]
+				i++
+			}
+		}
+		if i != len(nd.Values) {
+			return nil, fmt.Errorf("expected %d values, got %d", i, len(nd.Values))
+		}
+	} else if len(nd.Links) > 0 {
+		if expectLeaf {
+			return nil, errLeafExpected
+		}
+
+		for x := byte(0); x < internal.Width; x++ {
+			if nd.Bmap[x/8]&(1<<(x%8)) > 0 {
+				if i >= len(nd.Links) {
+					return nil, fmt.Errorf("expected at least %d links, found %d", i+1, len(nd.Links))
+				}
+				c := nd.Links[i]
+				if !c.Defined() {
+					return nil, errUndefinedCID
+				}
+				// TODO: check link hash function.
+				prefix := c.Prefix()
+				if prefix.Codec != cid.DagCBOR {
+					return nil, fmt.Errorf("internal amt nodes must be cbor, found %d", prefix.Codec)
+				}
+				n.links[x] = &link{cid: c}
+				i++
+			}
+		}
+		if i != len(nd.Links) {
+			return nil, fmt.Errorf("expected %d links, got %d", i, len(nd.Links))
+		}
+	} else if !allowEmpty {
+		return nil, errEmptyNode
+	}
+	return n, nil
+}
+
+func (nd *node) collapse(ctx context.Context, bs cbor.IpldStore, height int) (int, error) {
+	// If we have any links going "to the right", we can't collapse any
+	// more.
+	for _, l := range nd.links[1:] {
+		if l != nil {
+			return height, nil
+		}
+	}
+
+	// If we have _no_ links, we've collapsed everything.
+	if nd.links[0] == nil {
+		return 0, nil
```

Comment:
note: this is a bug in the current AMT. If we insert a high key, then delete it, we don't round-trip back to the empty AMT. (we fail to reset the height).